pkg/controllers/nodeclaim/underutilization/controller.go

```diff
@@ -0,0 +1,450 @@
+
+	if len(results.NewNodeClaims[0].NodeClaimTemplate.InstanceTypeOptions) == 0 {
+		if len(candidates) == 1 {
+			if len(incompatibleMinReqKey) > 0 {
+				c.recorder.Publish(disruptionevents.Unconsolidatable(candidates[0].Node, candidates[0].NodeClaim, fmt.Sprintf("minValues requirement is not met for %s", incompatibleMinReqKey))...)
+			} else {
+				c.recorder.Publish(disruptionevents.Unconsolidatable(candidates[0].Node, candidates[0].NodeClaim, "Can't replace with a cheaper node")...)
+			}
+		}
+		return false, nil
+	}
+	return true, nil
+}
+
+// propagates the consolidation decision to cluster state and also marks all
+// associated nodeclaims as underutilized
+func (c *Controller) setNodeClaimsUnderutilized(ctx context.Context, setTrue bool, commandID types.UID, candidates ...*state.StateNode) error {
+	for i := range candidates {
+		nc := candidates[i].NodeClaim
+		cond := nc.StatusConditions().GetCondition(v1beta1.Underutilized)
+		// If we want to set the condition and it's already true
+		// or if we don't want to set the condition and it isn't true
+		// e.g. (T && T) || (F && F), then do nothing
+		if (setTrue && cond.IsTrue()) || (!setTrue && !cond.IsTrue()) {
+			continue
+		}
+		deepCopy := nc.DeepCopy()
+		if setTrue {
+			nc.StatusConditions().SetCondition(apis.Condition{
+				Type:    v1beta1.Underutilized,
+				Status:  v1.ConditionTrue,
+				Message: (string)(commandID),
+			})
+		} else {
+			err := nc.StatusConditions().ClearCondition(v1beta1.Underutilized)
+			if err != nil {
+				return fmt.Errorf("clearing underutilized condition from nodeclaim %q", nc.Name)
+			}
+		}
+		if err := c.kubeClient.Status().Patch(ctx, nc, client.MergeFrom(deepCopy)); err != nil {
+			return fmt.Errorf("patching nodeclaim, %w", err)
+		}
+		logging.FromContext(ctx).With("nodeclaim", nc.Name).Debugf("%s nodeclaim underutilization status condition", lo.Ternary(setTrue, "set", "removed"))
+	}
+	return nil
+}
+
+// getCandidatePrices returns the sum of the prices of the given candidates
+func (c *Controller) getCandidatePrices(ctx context.Context, candidates ...*state.StateNode) (float64, error) {
+	c.refreshInstanceTypeMapping(ctx)
+	var price float64
+	for _, candidate := range candidates {
+		npName := candidate.Labels()[v1beta1.NodePoolLabelKey]
+		itName := candidate.Labels()[v1.LabelInstanceTypeStable]
+		ct := candidate.Labels()[v1beta1.CapacityTypeLabelKey]
+		zone := candidate.Labels()[v1.LabelTopologyZone]
+		offering, ok := c.instanceTypes[npName][itName].Offerings.Get(ct, zone)
+		if !ok {
+			return 0.0, fmt.Errorf("unable to determine offering for %s/%s/%s", itName, ct, zone)
+		}
+		price += offering.Price
+	}
+	return price, nil
+}
+
+// refreshInstanceTypeMapping tries to update the Controller's NodePoolToInstanceTypesMap
+// If the NodePool's instance types can't be discovered, the data is simply omitted.
+func (c *Controller) refreshInstanceTypeMapping(ctx context.Context) {
```

Comment:
What's the point of this being stored attached to the controller if we refresh this mapping every time? When I saw it attached to the controller state, I assumed that it was cached, but it looks like it isn't